torchgeo/models/changestar.py

```diff
@@ -0,0 +1,229 @@
+
+from typing import Any, Dict, List
+
+import torch
+from einops import rearrange
+from torch import Tensor
+from torch.nn.modules import (
+    BatchNorm2d,
+    Conv2d,
+    Identity,
+    Module,
+    ReLU,
+    Sequential,
+    UpsamplingBilinear2d,
+)
+
+from .farseg import FarSeg
+
+Identity.__module__ = "nn.Identity"
+
+
+class ChangeMixin(Module):
+    """This module enables any segmentation model to detect binary change.
+
+    If you use this model in your research, please cite the following paper:
+
+    * https://arxiv.org/abs/2108.07002
+    """
+
+    def __init__(
+        self,
+        in_channels: int = 128 * 2,
+        inner_channels: int = 16,
+        num_convs: int = 4,
+        scale_factor: float = 4.0,
+    ):
+        """Initializes a new ChangeMixin module.
+
+        Args:
+            in_channels: sum of channels of bitemporal feature maps
+            inner_channels: number of channels of inner feature maps
+            num_convs: number of convolution blocks
+            scale_factor: number of upsampling factor
+        """
+        super().__init__()  # type: ignore[no-untyped-call]
+        layers: List[Module] = [
+            Sequential(
+                Conv2d(in_channels, inner_channels, 3, 1, 1),
+                BatchNorm2d(inner_channels),  # type: ignore[no-untyped-call]
+                ReLU(True),
+            )
+        ]
+        layers += [
+            Sequential(
+                Conv2d(inner_channels, inner_channels, 3, 1, 1),
+                BatchNorm2d(inner_channels),  # type: ignore[no-untyped-call]
+                ReLU(True),
+            )
+            for _ in range(num_convs - 1)
+        ]
+
+        cls_layer = Conv2d(inner_channels, 1, 3, 1, 1)
+
+        layers.append(cls_layer)
+        layers.append(UpsamplingBilinear2d(scale_factor=scale_factor))
+
+        self.convs = Sequential(*layers)
+
+    def forward(self, bi_feature: Tensor) -> List[Tensor]:
+        """Forward pass of the model.
+
+        Args:
+            x: input bitemporal feature maps of shape [b, t, c, h, w]
+
+        Returns:
+            a list of bidirected output predictions
+        """
+        batch_size = bi_feature.size(0)
+        t1t2 = torch.cat(  # type: ignore[attr-defined]
+            [bi_feature[:, 0, :, :, :], bi_feature[:, 1, :, :, :]], dim=1
+        )
+        t2t1 = torch.cat(  # type: ignore[attr-defined]
+            [bi_feature[:, 1, :, :, :], bi_feature[:, 0, :, :, :]], dim=1
+        )
+
+        c1221 = self.convs(torch.cat([t1t2, t2t1], dim=0))  # type: ignore[attr-defined]
+        c12, c21 = torch.split(
+            c1221, batch_size, dim=0
+        )  # type: ignore[no-untyped-call]
+        return [c12, c21]
+
+
+class ChangeStar(Module):
+    """The base class of the network architecture of ChangeStar.
+
+    ChangeStar is composited of an any segmentation model and a ChangeMixin module.
+    This model is mainly used for binary/multi-class change detection under bitemporal
```

Comment:
Is it actually multi-class? Seems like in `ChangeMixin` you are hardcoding it to output a binary mask. Could we possibly add a `output_channels` arg to `ChangeMixin` to allow for multi-class change detection?

Reply:
Oh, Sorry. Only binary here.

Reply:
ChangeMixin is used to detect binary change, while the segmentation model provides semantic labels.
The semantic changes can be inferred by a binary change mask and two semantic masks.
Therefore, this network architecture can support binary and multi-class change detection.

Reply:
Ok that makes sense. It might be useful to add that somewhere in the docstring.

Reply:
Added.
Thanks for the suggestion.